File: 0118-pascals-triangle/0118-pascals-triangle.cpp
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        if (numRows <= 2) return numRows == 1? vector<vector<int>>{{1}} : vector<vector<int>>{{1}, {1, 1}};
        vector<vector<int>> dp{{{1}, {1, 1}}};
        for (int i = 3; i < numRows + 1; ++i){
            vector<int> curr(i, 1);
            for (int j = 1; j < i - 1; ++j){
                curr[j] = dp[i - 2][j - 1] + dp[i - 2][j];
            }
            dp.push_back(curr);
        }
        
        return dp;
    }
};

File: 0118-pascals-triangle/0118-pascals-triangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        if numRows == 1:
            return [[1]]

        if numRows == 2:
            return [[1], [1, 1]]

        ans = [[1], [1, 1]]
        for i in range(numRows - 2):
            curr = [1] * (len(ans[-1]) + 1)
            for j in range(1, len(ans[-1])):
                curr[j] = ans[-1][j - 1] + ans[-1][j]
            ans.append(curr)
        
        return ans